### PR TITLE
Feature/location view model

### PR DESCRIPTION
**Pull request by arthur-mrgt**
# PR: Integrate LocationViewModel and Update Location Tracking Mechanism

## Overview

This PR introduces the `LocationViewModel` class, which manages location updates and synchronizes the user’s current location with Firestore. It also updates `MainActivity` to integrate `LocationViewModel` with necessary dependencies, including `LocationService`, `LocationRepositoryFirestore`, and `PermissionManager`. Key enhancements include improved location tracking, permission handling, and Firestore updates for user location changes.

### 1. **LocationViewModel**  
- The `LocationViewModel` class encapsulates logic for managing location updates based on permissions.
- **Functions in `LocationViewModel`**:
  - `tryToStartLocationUpdates()`: Starts location updates if `ACCESS_FINE_LOCATION` permission is granted.
  - `stopLocationUpdates()`: Stops location updates and any permission-related listeners.
  - `observePermissions()`: Monitors `PermissionManager` for changes in `ACCESS_FINE_LOCATION` permission and adjusts location tracking accordingly.
  - `startServiceLocationUpdates()`: Initiates `LocationService` updates and updates Firestore with new locations.

- **Dependencies in `LocationViewModel`**:
  - `LocationService`: Handles low-level location requests and callbacks.
  - `LocationRepositoryFirestore`: Manages Firestore updates with user geolocation data.
  - `PermissionManager`: Manages location permission requests and state updates.

### 2. **MainActivity Integration**  
- Initializes `LocationViewModel` with dependencies on `LocationService`, `LocationRepositoryFirestore`, and `PermissionManager`.
- Monitors user login status to start or stop location updates automatically:
  - Starts location updates upon user login.
  - Stops updates on user logout.

### Location Update Flow

1. **Permission Check and Initialization**:
   - On `MainActivity` creation, `PermissionManager` initializes with `ACCESS_FINE_LOCATION`.
   - `LocationViewModel` calls `tryToStartLocationUpdates()` upon user login to check permission and initiate tracking if granted.

2. **Starting and Stopping Location Updates**:
   - If the permission is granted, `LocationViewModel` calls `startServiceLocationUpdates()`, which triggers `LocationService` to request location updates from `FusedLocationProviderClient`.
   - Each new location is processed by `LocationService`, which checks if it exceeds a predefined threshold distance to avoid unnecessary updates.
   - Valid locations are then passed back to `LocationViewModel`, which converts them into `GeoPoint` objects and calls `LocationRepositoryFirestore.setUserPosition()` to update Firestore.

3. **Permission Observation and Update Management**:
   - `LocationViewModel` listens to permission status changes via `PermissionManager`. If the `ACCESS_FINE_LOCATION` permission is revoked during active updates, it stops location updates immediately.
   - When the permission is re-granted, `LocationViewModel` attempts to restart location updates, ensuring permissions align with user settings and app requirements.

4. **Firestore Synchronization**:
   - `LocationRepositoryFirestore` handles the Firestore updates. When `LocationViewModel` sends a new `GeoPoint`, the repository calculates a high-precision geohash for optimized geolocation queries and updates the Firestore profile collection with the user’s geolocation data.

### Summary of Workflow

- **Initialization**: `MainActivity` initializes `LocationViewModel` and dependencies.
- **Login Event**: If logged in, `LocationViewModel` checks permissions and starts location tracking.
- **Location Update**: When a valid location is detected, it is updated in Firestore.
- **Permission Changes**: Observes permission changes to dynamically start or stop location updates.

---

This PR efficiently manages real-time user location tracking, responding dynamically to permission changes and user authentication status. It enhances user experience by ensuring location updates occur only when permissions and conditions are met, optimizing resource usage, and maintaining location data accuracy in Firestore.

### CODE REVIEW

File: app/src/main/java/com/github/se/icebreakrr/model/location/LocationViewModel.kt [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+   * or `false` if an error occurs.
+   */
+  private fun startServiceLocationUpdates() {
+    val locationUpdatesStarted =
+        locationService.startLocationUpdates(
+            onLocationUpdate = { location ->
+              val geoPoint = GeoPoint(location.latitude, location.longitude)
+              geoFirestoreRepository.setUserPosition(
+                  geoPoint) // Call Firestore update on location change
+            })
```
Comment (Kapestian) [inline app/src/main/java/com/github/se/icebreakrr/model/location/LocationViewModel.kt:89]: Nitpick: I see `startLocationUpdates` has a parameter "onError", shouldn't you use it here to handle errors more gracefully?

File: app/src/main/java/com/github/se/icebreakrr/MainActivity.kt [hunk 2]
```diff
@@ -36,24 +40,62 @@ import com.github.se.icebreakrr.ui.sections.FilterScreen
...
 
   override fun onCreate(savedInstanceState: Bundle?) {
     super.onCreate(savedInstanceState)
+
     // Retrieve the testing flag from the Intent
     val isTesting = intent?.getBooleanExtra("IS_TESTING", false) ?: false
 
-    requestNotificationPermission()
+    // Create and initialize the PermissionManager with the list of permissions required
+    requestNotificationPermission() // TODO remove this and use the PermissionManager instead
```
Comment (Kapestian) [inline app/src/main/java/com/github/se/icebreakrr/MainActivity.kt:67]: `requestNotificationPermission() // TODO remove this and use the PermissionManager instead`
Question: just to be sure, you commented that for Maximo to do it later, right?

**Review by Kapestian — COMMENTED** [review]
Dear Arthur, 
no wonder why people avoid reviews of your code: there is almost never anything to complain about!
Your code is of such high quality that it's almost annoying :)
I was just able to put two comments in the code, it's up to you to see if it's relevant or not. Also, take the time to check if logging statements are necessary or if you can remove them.
The only problem now is a bug with the tests (I think really the test file, not your source code). Tell me when it's fixed and I'll approve the PR.

Keep it up boss!

File: app/src/main/java/com/github/se/icebreakrr/MainActivity.kt (hunk 2 above)
Comment (arthur-mrgt) [reply app/src/main/java/com/github/se/icebreakrr/MainActivity.kt:80]: Yup exactly ;)

File: app/src/main/java/com/github/se/icebreakrr/model/location/LocationViewModel.kt (hunk 1 above)
Comment (arthur-mrgt) [reply app/src/main/java/com/github/se/icebreakrr/model/location/LocationViewModel.kt]: I've added a onError() parameter if we want to handle errors a different way in the future, for now just a error log is shown.

**Review by Kapestian — APPROVED** [review]
Your work is incredible, let's go!

**Pull request merged**